web-search-engines/langchain4j-web-search-engine-searchapi/src/main/java/dev/langchain4j/web/search/searchapi/SearchApiWebSearchEngine.java

```diff
@@ -0,0 +1,120 @@
+
+    private static final String BASE_URL = "https://www.searchapi.io";
+    private static final String DEFAULT_ENGINE = "google";
+
+    private final String apiKey;
+    private final String engine;
+    private final SearchApiClient client;
+    private final Map<String, Object> optionalParameters;
+
+    /**
+     * @param apiKey             the Search API key for accessing their API
+     * @param timeout            the timeout duration for API requests
+     *                           <p>
+     *                           Default value is 30 seconds.
+     * @param engine             the engine used by Search API to execute the search
+     *                           <p>
+     *                           Default engine is Google Search.
+     * @param optionalParameters optional parameters to be passed on every request of this the engine, they can be overridden by the WebSearchRequest additional parameters for matching keys
+     *                           <p>
+     *                           Check <a href="https://www.searchapi.io">Search API</a> for more information on available parameters for each engine
+     */
+    @Builder
+    public SearchApiWebSearchEngine(String apiKey,
+                                    Duration timeout,
+                                    String engine,
+                                    Map<String, Object> optionalParameters) {
+        this.apiKey = ensureNotBlank(apiKey, "apiKey");
+        this.engine = getOrDefault(engine, DEFAULT_ENGINE);
+        this.optionalParameters = getOrDefault(optionalParameters, new HashMap<>());
+        this.client = SearchApiClient.builder()
+                .baseUrl(BASE_URL)
+                .timeout(getOrDefault(timeout, ofSeconds(30)))
+                .build();
+    }
+
+    @Override
+    public WebSearchResults search(WebSearchRequest webSearchRequest) {
+        SearchApiWebSearchRequest request = SearchApiWebSearchRequest.builder()
+                .apiKey(apiKey)
+                .engine(engine)
+                .query(webSearchRequest.searchTerms())
+                .optionalParameters(optionalParameters)
+                .additionalRequestParameters(webSearchRequest.additionalParams())
+                .build();
+        SearchApiWebSearchResponse response = client.search(request);
+        return toWebSearchResults(response);
+    }
+
+    private WebSearchResults toWebSearchResults(SearchApiWebSearchResponse response) {
+        List<OrganicResult> organicResults = response.getOrganicResults();
+        Long totalResults = ((Integer) organicResults.size()).longValue(); // not ideal, but it may not be present in the response and is required not null by WebSearchInformationResult
+        WebSearchInformationResult searchInformation = WebSearchInformationResult.from(
+                totalResults,
+                getCurrentPage(response.getPagination()),
+                null
+        );
```

Comment:
Modify my suggestion if "total_results" not is got of this way.
```suggestion
        WebSearchInformationResult searchInformation = WebSearchInformationResult.from(
                Long.valueOf(getOrDefault(response.getSearchInformation().get("total_results"),0)),
                getCurrentPage(response.getPagination()),
                null
        );
```

Reply:
@mark-searchapi is `search_information->total_results` a common naming for every engine of SearchApi?

Reply:
Yes, `search_information->total_results` is a common name across all search engines. It might be missing (if the search engine for some reason does not return it), but that does not mean there are no results.